### ISSUE TITLE
[JENKINS-10060] cannot use the "pass-through git commit that was built" parameter with triggered jobs

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#2031]
<p>I am trying to distribute my test suite jobs to run in parallel by kicking them off from a "master" build job which triggers other jobs. I want all test suites to run against a consistent state of my repository, so I added the  "pass-through git commit that was built" parameter. But this causes all sub-jobs to fail with the error message below. This is blocking me from setting up jenkins jobs the way I want.</p>

<p>I suppose this could be a misconfiguration. But my sub-jobs run fine when they are run individually, and I don't see any git parameter I can add in my sub-job configuration. Ideally, I want the "pass-through git commit" feature to just work without any special re-configuration.</p>

<p>This is related to issue <a href="https://issues.jenkins.io/browse/JENKINS-9049" title="NPE while initializing Git plugin" class="issue-link" data-issue-key="JENKINS-9049"><del>JENKINS-9049</del></a>, but that issue's exception is different and it's related to the Gerrit plugin (which I am not using), so I thought I should open a new issue.</p>

<p>The exception is:</p>

<p>Started by upstream project "Master" build number 5<br/>
Building on master<br/>
Checkout:workspace / /data/jenkins/jobs/Master-unit_tests/workspace - hudson.remoting.LocalChannel@​126be4cc<br/>
Using strategy: Default<br/>
Last Built Revision: Revision 70532721cd3016b4746da4c89a9b7cc85a11bae4 (origin/master)<br/>
Checkout:workspace / /data/jenkins/jobs/Master-unit_tests/workspace - hudson.remoting.LocalChannel@​126be4cc<br/>
Fetching changes from the remote Git repository<br/>
Fetching upstream changes from git@​git.example.com:my_project.git<br/>
Commencing build of Revision e4d1ff0056235b827934b1d9a0352a6a8f04e286 (detached)<br/>
Checking out Revision e4d1ff0056235b827934b1d9a0352a6a8f04e286 (detached)<br/>
FATAL: no remote from branch name (detached)<br/>
hudson.plugins.git.GitException: no remote from branch name (detached)<br/>
	at hudson.plugins.git.GitAPI.setupSubmoduleUrls(GitAPI.java:667)<br/>
	at hudson.plugins.git.GitSCM$4.invoke(GitSCM.java:1158)<br/>
	at hudson.plugins.git.GitSCM$4.invoke(GitSCM.java:1120)<br/>
	at hudson.FilePath.act(FilePath.java:758)<br/>
	at hudson.FilePath.act(FilePath.java:740)<br/>
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1120)<br/>
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1181)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.checkout(AbstractBuild.java:536)<br/>
	at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:424)<br/>
	at hudson.model.Run.run(Run.java:1375)<br/>
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)<br/>
	at hudson.model.ResourceController.execute(ResourceController.java:88)<br/>
	at hudson.model.Executor.run(Executor.java:145)</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amurray.jpg" title="amurray's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amurray">amurray</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060" target="_blank">cannot use the "pass-through git commit that was built" parameter with triggered jobs</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-08-03T03:10:14+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I am trying to distribute my test suite jobs to run in parallel by kicking them off from a "master" build job which triggers other jobs. I want all test suites to run against a consistent state of my repository, so I added the  "pass-through git commit that was built" parameter. But this causes all sub-jobs to fail with the error message below. This is blocking me from setting up jenkins jobs the way I want.</p>

<p>I suppose this could be a misconfiguration. But my sub-jobs run fine when they are run individually, and I don't see any git parameter I can add in my sub-job configuration. Ideally, I want the "pass-through git commit" feature to just work without any special re-configuration.</p>

<p>This is related to issue <a href="https://issues.jenkins.io/browse/JENKINS-9049" title="NPE while initializing Git plugin" class="issue-link" data-issue-key="JENKINS-9049"><del>JENKINS-9049</del></a>, but that issue's exception is different and it's related to the Gerrit plugin (which I am not using), so I thought I should open a new issue.</p>

<p>The exception is:</p>

<p>Started by upstream project "Master" build number 5
Building on master
Checkout:workspace / /data/jenkins/jobs/Master-unit_tests/workspace - hudson.remoting.LocalChannel@126be4cc
Using strategy: Default
Last Built Revision: Revision 70532721cd3016b4746da4c89a9b7cc85a11bae4 (origin/master)
Checkout:workspace / /data/jenkins/jobs/Master-unit_tests/workspace - hudson.remoting.LocalChannel@126be4cc
Fetching changes from the remote Git repository
Fetching upstream changes from git@git.example.com:my_project.git
Commencing build of Revision e4d1ff0056235b827934b1d9a0352a6a8f04e286 (detached)
Checking out Revision e4d1ff0056235b827934b1d9a0352a6a8f04e286 (detached)
FATAL: no remote from branch name (detached)
hudson.plugins.git.GitException: no remote from branch name (detached)
	at hudson.plugins.git.GitAPI.setupSubmoduleUrls(GitAPI.java:667)
	at hudson.plugins.git.GitSCM$4.invoke(GitSCM.java:1158)
	at hudson.plugins.git.GitSCM$4.invoke(GitSCM.java:1120)
	at hudson.FilePath.act(FilePath.java:758)
	at hudson.FilePath.act(FilePath.java:740)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1120)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1181)
	at hudson.model.AbstractBuild$AbstractRunner.checkout(AbstractBuild.java:536)
	at hudson.model.AbstractBuild$AbstractRunner.run(AbstractBuild.java:424)
	at hudson.model.Run.run(Run.java:1375)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
	at hudson.model.ResourceController.execute(ResourceController.java:88)
	at hudson.model.Executor.run(Executor.java:145)</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>64-bit linux</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151213&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151213">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Please report the version of the Git plugin.</p>
</pre>
</details>

</ul>
</details>

<p>Please report the version of the Git plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amurray.jpg" title="amurray's avatar" /> amurray:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151214&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151214">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Version 1.1.9. </p>

<p>I see that 1.1.11 is now available. Is this perhaps fixed in the latest version? I will try it soon...</p>
</pre>
</details>

</ul>
</details>

<p>Version 1.1.9. </p>

<p>I see that 1.1.11 is now available. Is this perhaps fixed in the latest version? I will try it soon...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151216&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151216">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>OK, I figured out what's going on.</p>

<p>Git plugin tries to be clever about submodule handling &#8212; by which I mean we try to figure out a "better" location to check out modules.</p>

<p>Imagine the situation where you have something like tree-structured repositories: at the root you have corporate central repositories that you ultimately push to, which all .gitmodules point to, then you have intermediate team local repository, which is assumed to be a non-bare repository (say, a checked out copy on a shared server accessed via SSH.) Maybe this is where you are pushing, so that you can have the changes tested by Jenkins before it hits the central. Or maybe this is where you do merge test with others.</p>

<p>In anticipation of this use case, Git plugin looks at which remote repository the commit came from, and if it looks like a non-bare repository (meaning if it slooks like the team local repository above), it tries to update modules from subdirectories of this repository, assuming that it's a fully populated checked out copy.</p>

<p>Note that there's a lot of heuristics and assumptions involved in this. And this code wasn't handling the case correctly if we are building a specific SHA1 (either via user-specified value, or via pass-through Git commit.) So the fix involves in making this logic a bit more robust.</p>

<p>I have to say that, for my personal taste, this feels like too much heuristics from such unreliable clues, but such are the current behaviour, and we need to preserve this.</p>

<p>I'm testing the fix now.</p>
</pre>
</details>

</ul>
</details>

<p>OK, I figured out what's going on.</p>

<p>Git plugin tries to be clever about submodule handling &#8212; by which I mean we try to figure out a "better" location to check out modules.</p>

<p>Imagine the situation where you have something like tree-structured repositories: at the root you have corporate central repositories that you ultimately push to, which all .gitmodules point to, then you have intermediate team local repository, which is assumed to be a non-bare repository (say, a checked out copy on a shared server accessed via SSH.) Maybe this is where you are pushing, so that you can have the changes tested by Jenkins before it hits the central. Or maybe this is where you do merge test with others.</p>

<p>In anticipation of this use case, Git plugin looks at which remote repository the commit came from, and if it looks like a non-bare repository (meaning if it slooks like the team local repository above), it tries to update modules from subdirectories of this repository, assuming that it's a fully populated checked out copy.</p>

<p>Note that there's a lot of heuristics and assumptions involved in this. And this code wasn't handling the case correctly if we are building a specific SHA1 (either via user-specified value, or via pass-through Git commit.) So the fix involves in making this logic a bit more robust.</p>

<p>I have to say that, for my personal taste, this feels like too much heuristics from such unreliable clues, but such are the current behaviour, and we need to preserve this.</p>

<p>I'm testing the fix now.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151217&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151217">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 src/test/java/hudson/plugins/git/GitSCMTest.java
<a href="http://jenkins-ci.org/commit/git-plugin/cefbb29df219a2bd9742e8e2dcb687988e56a133" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/cefbb29df219a2bd9742e8e2dcb687988e56a133</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-10060" title="cannot use the &quot;pass-through git commit that was built&quot; parameter with triggered jobs" class="issue-link" data-issue-key="JENKINS-10060"><del>JENKINS-10060</del></a> added a regression test case.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 src/test/java/hudson/plugins/git/GitSCMTest.java<br/>
<a href="http://jenkins-ci.org/commit/git-plugin/cefbb29df219a2bd9742e8e2dcb687988e56a133" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/cefbb29df219a2bd9742e8e2dcb687988e56a133</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-10060" title="cannot use the "pass-through git commit that was built" parameter with triggered jobs" class="issue-link" data-issue-key="JENKINS-10060"><del>JENKINS-10060</del></a> added a regression test case.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151218&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151218">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 src/main/java/hudson/plugins/git/GitAPI.java
<a href="http://jenkins-ci.org/commit/git-plugin/c8c7c0afb6f997ab4b743471f49c520875e52e7e" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/c8c7c0afb6f997ab4b743471f49c520875e52e7e</a>
Log:
  <span class="error">&#91;FIXED JENKINS-10060&#93;</span> made the handling more robust.</p>

<p>See the comment in the code and in the ticket for discussion of what was going wrong and why I think this fix is a reasonable one.</p>


<p>Compare: <a href="https://github.com/jenkinsci/git-plugin/compare/7219d74...c8c7c0a" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/compare/7219d74...c8c7c0a</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 src/main/java/hudson/plugins/git/GitAPI.java<br/>
<a href="http://jenkins-ci.org/commit/git-plugin/c8c7c0afb6f997ab4b743471f49c520875e52e7e" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/c8c7c0afb6f997ab4b743471f49c520875e52e7e</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-10060&#93;</span> made the handling more robust.</p>

<p>See the comment in the code and in the ticket for discussion of what was going wrong and why I think this fix is a reasonable one.</p>


<p>Compare: <a href="https://github.com/jenkinsci/git-plugin/compare/7219d74...c8c7c0a" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/compare/7219d74...c8c7c0a</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-10060?focusedId=151220&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-151220">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/plugins_git-plugin/207/" class="external-link" target="_blank" rel="nofollow noopener">plugins_git-plugin #207</a>
     <a href="https://issues.jenkins.io/browse/JENKINS-10060" title="cannot use the &quot;pass-through git commit that was built&quot; parameter with triggered jobs" class="issue-link" data-issue-key="JENKINS-10060"><del>JENKINS-10060</del></a> added a regression test case.
<span class="error">&#91;FIXED JENKINS-10060&#93;</span> made the handling more robust.</p>

<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/git-plugin/commit/cefbb29df219a2bd9742e8e2dcb687988e56a133" class="external-link" target="_blank" rel="nofollow noopener">cefbb29df219a2bd9742e8e2dcb687988e56a133</a>
Files : </p>
<ul>
	<li>src/test/java/hudson/plugins/git/GitSCMTest.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/git-plugin/commit/c8c7c0afb6f997ab4b743471f49c520875e52e7e" class="external-link" target="_blank" rel="nofollow noopener">c8c7c0afb6f997ab4b743471f49c520875e52e7e</a>
Files : </p>
<ul>
	<li>src/main/java/hudson/plugins/git/GitAPI.java</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/plugins_git-plugin/207/" class="external-link" target="_blank" rel="nofollow noopener">plugins_git-plugin #207</a><br/>
     <a href="https://issues.jenkins.io/browse/JENKINS-10060" title="cannot use the "pass-through git commit that was built" parameter with triggered jobs" class="issue-link" data-issue-key="JENKINS-10060"><del>JENKINS-10060</del></a> added a regression test case.<br/>
<span class="error">&#91;FIXED JENKINS-10060&#93;</span> made the handling more robust.</p>

<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/git-plugin/commit/cefbb29df219a2bd9742e8e2dcb687988e56a133" class="external-link" target="_blank" rel="nofollow noopener">cefbb29df219a2bd9742e8e2dcb687988e56a133</a><br/>
Files : </p>
<ul>
	<li>src/test/java/hudson/plugins/git/GitSCMTest.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/git-plugin/commit/c8c7c0afb6f997ab4b743471f49c520875e52e7e" class="external-link" target="_blank" rel="nofollow noopener">c8c7c0afb6f997ab4b743471f49c520875e52e7e</a><br/>
Files : </p>
<ul>
	<li>src/main/java/hudson/plugins/git/GitAPI.java</li>
</ul>
